opengrok-web/src/main/java/org/opengrok/web/api/ApiTask.java

```diff
@@ -50,68 +55,145 @@
 
     private final Response.Status responseStatus;
 
+    private Future<Object> future;
+
+    private final Map<Class<?>, Response.Status> exceptionStatusMap = new HashMap<>();
+
     /**
      * @param path request path (for identification)
-     * @param runnable Runnable object
+     * @param callable Callable object
      */
-    public ApiTask(String path, Runnable runnable) {
-        this(path, runnable, Response.Status.OK);
+    public ApiTask(String path, Callable<Object> callable) {
+        this(path, callable, Response.Status.OK);
     }
 
     /**
      * @param path request path (for identification)
-     * @param runnable Runnable object
+     * @param callable Callable object
      * @param status request status to return after the runnable is done
      */
-    public ApiTask(String path, Runnable runnable, Response.Status status) {
-        this.runnable = runnable;
+    public ApiTask(String path, Callable<Object> callable, Response.Status status) {
+        this(path, callable, status, null);
+    }
+
+   /**
+    * @param path request path (for identification)
+    * @param callable Callable object
+    * @param status request status to return after the runnable is done
+    * @param exceptionStatusMap map of {@link Exception} to {@link Response.Status}
+    */
+    public ApiTask(String path, Callable<Object> callable, Response.Status status,
+                   Map<Class<?>, Response.Status> exceptionStatusMap) {
+        this.callable = callable;
         this.uuid = UUID.randomUUID();
         this.responseStatus = status;
         this.path = path;
         this.state = ApiTaskState.INITIAL;
+        if (exceptionStatusMap != null) {
+            this.exceptionStatusMap.putAll(exceptionStatusMap);
+        }
     }
 
+    /**
+     * The UUID is randomly generated in the constructor.
+     * @return UUID
+     */
     public UUID getUuid() {
         return uuid;
     }
 
     /**
-     * @return response status
+     * @return response status to be used when the task was successfully completed
      */
-    public Response.Status getResponseStatus() {
+    Response.Status getResponseStatus() {
         return responseStatus;
     }
 
     /**
      * Set status as submitted.
      */
-    public void setSubmitted() {
+    void setSubmitted() {
         state = ApiTaskState.SUBMITTED;
     }
 
+    /**
+     * @return whether the API task successfully completed
+     */
     public boolean isCompleted() {
         return state.equals(ApiTaskState.COMPLETED);
     }
 
-    public void setCompleted() {
+    void setCompleted() {
         state = ApiTaskState.COMPLETED;
     }
 
+    /**
+     * @param future Future object used for tracking the progress of the API task
+     */
+    public void setFuture(Future<Object> future) {
+        this.future = future;
+    }
+
+    /**
+     * @return whether the task is finished (normally or with exception)
+     */
+    public boolean isDone() {
```

Comment:
Why this? Isn't `isCompleted` an indicator already?

Reply:
I was a bit torn whether not to turn `isCompleted()` to `isDone()` however decided to leave it as it is quite handy for testing.